### PR TITLE
[Feature] Support MiniMax-M3 Wint4Abf16 and Win4Afp8 on hopper.

### PR DESCRIPTION
**Pull request by yuyu5333**
## Motivation

Upstream https://github.com/sgl-project/sglang/pull/25763, https://github.com/sgl-project/sglang/pull/27944, https://github.com/sgl-project/sglang/pull/28715

This PR  supports Wint4 inference for the MiniMax-M3 model, and the Wint4 model is quantized from https://huggingface.co/MiniMaxAI/MiniMax-M3.

> **This PR is a rebased version of https://github.com/sgl-project/sglang/pull/29107**

## Accuracy Tests

Deployment with TP8:
``` bash
GLOO_SOCKET_IFNAME=eth0 \
NCCL_MIN_NCHANNELS=24 \
NCCL_IB_QPS_PER_CONNECTION=8 \
python -m sglang.launch_server \
  --trust-remote-code \
  --model-path /data00/nvme1/MiniMax-M3-W4A16 \
  --reasoning-parser auto \
  --tool-call-parser auto \
  --tp 8 \
  --mem-fraction-static 0.75 \
  --chunked-prefill-size 8192 \
  --watchdog-timeout 3600 \
  --host 0.0.0.0 \
  --port 30000
```

<meta charset="utf-8"><div data-page-id="B3kHdQchGoK9XBxpDoDcg3oUndg" data-lark-html-role="root" data-docx-has-block-data="true"><div>
 Acc. | W4ABF16 | W4AFP8
-- | -- | --
GSM8K | 0.975 | 0.976
MMLU | 0.841 | 0.841
GPQA Diamond | 0.969 | 0.969
Aime25 pass@1 | 90.23% | 91.80%

</div></div>

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29799310603](https://github.com/sgl-project/sglang/actions/runs/29799310603)
Latest PR Test (Extra): :x: [Run #29799310490](https://github.com/sgl-project/sglang/actions/runs/29799310490)

### CODE REVIEW

File: python/sglang/srt/layers/moe/cutlass_w4a8_moe.py [hunk 1]
```diff
@@ -198,7 +200,36 @@ def cutlass_w4a8_moe(
...
+        )
+
+        alpha = float(gemm1_alpha)
+        limit = None if gemm1_clamp_limit is None else float(gemm1_clamp_limit)
+        intermediate = torch.empty((m * topk, n), dtype=torch.bfloat16, device=device)
+        swiglu_gpt_oss_sigmoid_alpha_contiguous(intermediate, c1, alpha, limit)
+        if a2_scale is None:
+            a2_scale = torch.zeros(1, dtype=torch.float32, device=device)
+            # is_static=False: kernel computes absmax-based scale itself.
+            per_tensor_quant_fp8(intermediate, intermediate_q, a2_scale, False)
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/moe/cutlass_w4a8_moe.py:228]: `c1` is allocated with `torch.empty`, and under EP the sentinel routes are excluded from GEMM, so this full-tensor activation/absmax reads unwritten rows and can inflate `a2_scale`; make the SwiGLU-OAI quantization respect `expert_offsets[-1]` as the SiLU path below does, and add an EP case with dropped routes.

File: python/sglang/srt/models/minimax_m3.py [hunk 2]
```diff
@@ -1468,6 +1468,24 @@ def determine_num_fused_shared_experts(self):
             disable_reason = "Shared experts fusion is not supported together with expert parallelism yet."
         elif get_moe_a2a_backend().is_deepep():
             disable_reason = "Shared experts fusion is not supported when Deepep MoE backend is enabled."
+        elif self.quant_config is not None and any(
+            "shared_experts" in pat
+            for pat in (getattr(self.quant_config, "ignore", []) or [])
+        ):
```
Comment (JustinTong0323) [inline python/sglang/srt/models/minimax_m3.py:1474]: This substring check does not implement compressed-tensors' supported `re:` semantics (for example, `re:.*shared.*` matches the layer but misses this branch), so fusion can remain enabled for an ignored BF16 shared expert; use `should_ignore_layer` on the real shared-expert FQN, apply the same fix to the VL path, and cover exact/regex patterns.

File: python/sglang/srt/layers/moe/fused_moe_triton/fused_marlin_moe.py [hunk 3]
```diff
@@ -55,14 +55,29 @@ def swiglu_limit_func(
...
+    if gemm1_limit is not None and gemm1_limit > 0:
+        gate = torch.clamp(gate, max=gemm1_limit)
+        up = torch.clamp(up, min=-gemm1_limit, max=gemm1_limit)
+    out_dtype = output.dtype
+    activated = (
+        gate.to(torch.float32)
+        * torch.sigmoid(gemm1_alpha * gate.to(torch.float32))
+        * (up.to(torch.float32) + 1.0)
+    )
+    output.copy_(activated.to(out_dtype))
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/moe/fused_moe_triton/fused_marlin_moe.py:80]: This shared GPT-OSS helper now promotes every activation to FP32, silently changing existing GPT-OSS Marlin numerics and adding hot-path memory traffic; keep its old behavior and add a MiniMax-specific helper, or include GPT-OSS logits/e2e and performance evidence for the deliberate change.

**Review by JustinTong0323 — COMMENTED** [review]
Current-head Hopper jobs are skipped because the `run-ci` label is missing; after fixing the inline issues, run H100/H200 coverage and report stop-rate/finish_reason alongside accuracy.
